File: K-Nearest Neighbours/Qn-2.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
# 1 - Data Loding
zoo=pd. read_csv('C:\\Users\\Biswa\\Desktop\\360DigiTMG\\18 . Module 18 - Machine Learning K-Nearest Neighbour\\Assignment\\Dataset\\zoo.csv')

zoo.head()
zoo.shape
zoo.columns
zoo.info()
zoo.isnull().sum()
zoo.describe()

# 2- EDA

# Type wise animal catSegory plots
zoo_eda=zoo.drop('animal name',axis=1)

for i in zoo_eda.columns:
    fig=plt.subplots(figsize=(12,6))
    sns.countplot(zoo_eda['type'],hue=zoo_eda[i])
    plt.legend(loc='upper right')
    plt.title(i)
    plt.show()

for i in zoo_eda.columns:
    g=sns.catplot(i,col='type',data=zoo_eda,kind='count',col_wrap=7,height=2.5,aspect=0.8)
    plt.show()
    
 # Corelation Heatmap
cor=zoo.corr()  
figure=plt.subplots(figsize=(12,8))
sns.heatmap(cor,xticklabels='auto',yticklabels='auto',cmap='coolwarm',annot=True)
plt.show()

#Segegating features & target columns

x=zoo.iloc[:,1:-1]
y=zoo.iloc[:,-1]

# Vif Values
from statsmodels.stats.outliers_influence import variance_inflation_factor

vif=[variance_inflation_factor(zoo_eda.values,i) for i in range(zoo_eda.shape[1])]
vif=pd.DataFrame({'VIF':vif},index=zoo_eda.columns)
vif=vif.reset_index()

vif_score=sns.catplot(x='VIF',y='index',kind='bar',data=vif,aspect=2)
(vif_score.set_titles('Vif Score'))

# Feature Importance
from sklearn.ensemble import ExtraTreesClassifier

model=ExtraTreesClassifier()
model.fit(x,y)
feature_importance=pd.Series(model.feature_importances_,index=x.columns)
feature_importance=pd.DataFrame(feature_importance)
feature_importance.columns=['Importance']
feature_importance=feature_importance.reset_index()

feature_score=sns.catplot(y='index',x='Importance',data=feature_importance,kind='bar',aspect=2)
plt.show()

x_new=zoo_eda.drop(columns=['predator','venomous','domestic','catsize'])

# Modeling:
    
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score

accuracy_score=[]
for i in range(1,40,2):
    knn=KNeighborsClassifier(n_neighbors=i)
    score=cross_val_score(knn,x_new,y,cv=10)
    accuracy_score.append(score.mean())    

f,ax=plt.subplots(figsize=(12,6)) 
plt.plot(range(1,40,2),accuracy_score,marker='o')
plt.xlabel('K Number')
plt.ylabel('Accuracy')
plt.title('Accuracy rate')
plt.show()

# Modeling with k=3

knn1=KNeighborsClassifier(n_neighbors=3)
score1=cross_val_score(knn1,x_new,y,cv=10)

score1.mean()
score1.max()
score1.min()

# Confusion matrix & Classificatoin report
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=0)

from sklearn.metrics import confusion_matrix,classification_report

knn2=KNeighborsClassifier(n_neighbors=3)
knn2.fit(x_train,y_train)

y_pred=knn2.predict(x_test)

cm=confusion_matrix(y_test,y_pred)
report=classification_report(y_test,y_pred)
print(report)

File: K-Nearest Neighbours/Qn-2.R
library(class)
library(gmodels)
library(caret)
zoo<-read.csv('C:\\Users\\Biswa\\Desktop\\360DigiTMG\\18 . Module 18 - Machine Learning K-Nearest Neighbour\\Assignment\\Dataset\\zoo.csv')

colnames(zoo)
head(zoo)

#Copying our data for knn
data_class<-zoo

type<-zoo[,c(17:18)]
type$catsize<-NULL
data_class<-zoo[-18]
data_class$animal.name<-NULL

data_class=scale(data_class)
#Segregating our data into input & output sets
set.seed(1234)

smp_size=floor(0.75*nrow(data_class))
train_ind<-sample(nrow(data_class),size=smp_size)

#Creating test & train set
class_pred_train<-data_class[train_ind,]
class_pred_test<-data_class[-train_ind,]

type_out_train<-type[train_ind,]
type_out_test<-type[-train_ind,]

type_pred_knn<-knn(train=class_pred_train,test=class_pred_test,cl=type_out_train,k=10)

#Model Evaluation
type_out_test<-data.frame(type_out_test)

class_comparison<-data.frame(type_pred_knn,type_out_test)

names(class_comparison)<-c('Predicted Type','Observed Type')

head(class_comparison)

#Creating Table
CrossTable(x=class_comparison$`Observed Type`,y=class_comparison$`Predicted Type`,prop.chisq = FALSE,prop.c = FALSE,prop.r = FALSE,prop.t = FALSE)

#Using caret package for finding best k value
type_pred_caret<-train(class_pred_train,type_out_train,method = 'knn',preProcess = c('center','scale'))
type_pred_caret

plot(type_pred_caret)

#Confusion Martix

knnPredict<-predict(type_pred_caret,newdata = class_pred_test)

library(e1071)
confusionMatrix(factor(knnPredict,levels=1:7),factor(type_out_test$type_out_test,levels=1:7))

#Fianl Model wth K=9
type_pred_knn<-knn(train=class_pred_train,test=class_pred_test,cl=type_out_train,k=9)

#Model Evaluation
type_out_test<-data.frame(type_out_test)

class_comparison<-data.frame(type_pred_knn,type_out_test)

names(class_comparison)<-c('Predicted Type','Observed Type')

head(class_comparison)

#Creating Table
CrossTable(x=class_comparison$`Observed Type`,y=class_comparison$`Predicted Type`,prop.chisq = FALSE,prop.c = FALSE,prop.r = FALSE,prop.t = FALSE)

#Confusion Martix

knnPredict<-predict(type_pred_caret,newdata = class_pred_test)

library(e1071)
confusionMatrix(factor(knnPredict,levels=1:7),factor(type_out_test$type_out_test,levels=1:7))








